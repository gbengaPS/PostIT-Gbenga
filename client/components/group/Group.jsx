import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import swal from 'sweetalert';
import {
  getGroupMessages,
  getGroupMembers,
  leaveGroup,
  leaveGroupSuccess } from '../../actions/groupActions';
import Messages from '../message/Messages';
import GroupMembers from './GroupMembers';

/**
 * @class Group
 *
 * @extends {React.Component}
 */
export class Group extends React.Component {

  /**
   * @description Creates an instance of Group.
   *
   * @param {object} props -prop object
   *
   * @memberof Group
   */
  constructor(props) {
    super(props);
    this.leaveGroup = this.leaveGroup.bind(this);
  }

  /**
   *
   * @returns {void} -returns nothing
   */
  componentDidMount() {
    const { groupId } = this.props.match.params;
    this.props.getMessages(groupId, this.props.history);
    this.props.getGroupMembers(groupId);
  }

  /**
   *
   * @returns {void}
   */
  componentWillUpdate() {
    $('.dropdown-button').dropdown();
    $('select').material_select();
  }
  /**
   * @description leave group function
   *
   * @returns {void} -returns nothing
   */
  leaveGroup() {
    swal({
      title: 'Leave group warning',
      text: 'Are you sure you want to leave group?',
      icon: 'warning',
      buttons: ['Cancel', 'Yes'],
      dangerMode: true,
    })
    .then((leave) => {
      if (leave) {
        const { groupId } = this.props.match.params;
        this.props.leaveGroup(groupId);
      }
    });
  }

  /**
   * @description render function
   *
   * @returns {jsx} -jsx representation of the component
   */
  render() {
    if (this.props.leftGroup) {
      swal('you left group')
      .then(() => {
        this.props.leaveGroupSuccess(false);
        this.props.history.push('/dashboard');
      });
    }
    const { groupId } = this.props.match.params;
    let numberOfGroupMembers;
    let groupName;
    if (this.props.groupMembers.length > 0) {
      groupName = this.props.currentGroup.groupName;
      numberOfGroupMembers = this.props.groupMembers.length;
    }
    return (
      <div>
        <div>
        <div className="col s12 m7 component-container">
          <div className="dashboard-header">
          <Link
          to=""
          id="more-vert"
          className="right dropdown-button"
          data-activates="group-more"
          >
          <i className="material-icons">more_vert</i>
          </Link>
          <span className="right">
          <i className="material-icons">person</i>
          <span className="group-members">
            {numberOfGroupMembers}
          </span>
          </span>
          <span className="bold big">
            {groupName}
          </span>
          </div>
        </div>
            <ul id="group-more" className="dropdown-content">
              <li>
                <Link
                to={`/group/${groupId}/addmembers`}
                >
                Add Members
                </Link>
              </li>
              <li>
                <Link
                to="#"
                id="leave-group"
                onClick={this.leaveGroup}
                >
                Leave group
                </Link>
              </li>
            </ul>
          <div className="col s12 m7 component-container">
            <Messages groupId={this.props.match.params.groupId} />
          </div>
          <GroupMembers />
        </div>
      </div>
    );
  }
}
Group.propTypes = {
  getGroupMembers: PropTypes.func.isRequired,
  getMessages: PropTypes.func.isRequired,
  groupMembers: PropTypes.array,
  match: PropTypes.object,
  leaveGroup: PropTypes.func,
  history: PropTypes.object,
  leftGroup: PropTypes.bool,
  leaveGroupSuccess: PropTypes.func.isRequired,
  currentGroup: PropTypes.object,
};
/**
 * @description Maps state to props
 *
 * @param {object} state -application state
 *
 * @returns {object} -returns part of the state
*/
const mapStateToProps = state => (
  {
    leftGroup: state.groupReducer.leftGroup,
    groupMembers: state.groupReducer.members,
    currentGroup: state.groupReducer.currentGroup,
  }
);

/**
 * @description Maps dispatch to props
 *
 * @param {function} dispatch -dispatch function
 *
 * @returns {object} -actions to be dispatched
 */
const mapDispatchToProps = dispatch => (
  {
    getMessages: (groupId, history) => {
      dispatch(getGroupMessages(groupId, history));
    },
    getGroupMembers: (groupId) => {
      dispatch(getGroupMembers(groupId));
    },
    leaveGroup: (groupId) => {
      dispatch(leaveGroup(groupId));
    },
    leaveGroupSuccess: (leftGroup) => {
      dispatch(leaveGroupSuccess(leftGroup));
    }
  }
);

export default connect(mapStateToProps, mapDispatchToProps)(Group);
