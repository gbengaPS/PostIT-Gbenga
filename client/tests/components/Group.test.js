/* eslint-disable prettier */
import React from 'react';
import 'mock-local-storage';
import Group from '../../components/group/Group';

global.window = {};

window.localStorage = global.localStorage;
localStorage.setItem('postitUser', JSON.stringify({}));
describe('Group component', () => {
  const props = {
    getGroupMembers: () => {},
    getMessages: () => {},
    groupMembers: [],
    leaveGroup: () => {},
    history: {},
    leftGroup: false,
    leaveGroupSuccess: () => {},
    match: { params: { groupId: 1 } },
    currentGroup: {}
  };
  const wrapper = shallow(<Group.WrappedComponent {...props} />);
  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('should have componentWillUpdate method', () => {
    const spy = jest.spyOn(wrapper.instance(), 'componentWillUpdate');
    wrapper.setProps({ leftGroup: true });
    wrapper.instance().componentWillUpdate();
    expect(spy).toBeDefined();
  });
  it('should have a leave group function', () => {
    const leaveGroupSpy = jest.spyOn(wrapper.instance(), 'leaveGroup');
    wrapper.instance().leaveGroup();
    expect(leaveGroupSpy).toHaveBeenCalled();
  });
});
