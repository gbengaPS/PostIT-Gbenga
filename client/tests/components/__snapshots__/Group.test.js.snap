// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Group component should render correctly 1`] = `
<div>
  <div>
    <div
      className="col s12 m7 component-container"
    >
      <div
        className="dashboard-header"
      >
        <Link
          className="right dropdown-button"
          data-activates="group-more"
          id="more-vert"
          replace={false}
          to=""
        >
          <i
            className="material-icons"
          >
            more_vert
          </i>
        </Link>
        <span
          className="right"
        >
          <i
            className="material-icons"
          >
            person
          </i>
          <span
            className="group-members"
          />
        </span>
        <span
          className="bold big"
        />
      </div>
    </div>
    <ul
      className="dropdown-content"
      id="group-more"
    >
      <li>
        <Link
          replace={false}
          to="/group/1/addmembers"
        >
          Add Members
        </Link>
      </li>
      <li>
        <Link
          id="leave-group"
          onClick={[Function]}
          replace={false}
          to="#"
        >
          Leave group
        </Link>
      </li>
      <li>
        <Link
          id="leave-group"
          onClick={[Function]}
          replace={false}
          to="#"
        >
          Delete Group
        </Link>
      </li>
    </ul>
    <div
      className="col s12 m7 component-container"
    >
      <Connect(Messages)
        groupId={1}
      />
    </div>
    <Connect(GroupMembers) />
  </div>
</div>
`;
